# Ecological Genomics Nothebook

## Author: Jorge A. Ruiz-Arocho
### Affiliation: University of Vermont, PSS  
### E-mail contact: jorge.ruizarocho@uvm.edu


### Start Date: 2020-01-13
### End Date: 2020-05-08
### Project Descriptions:   





# Table of Contents:   
* [Entry 1: 2020-01-13, Introductory Class](#id-section1)
* [Entry 2: 2020-01-15, First Info Updates](#id-section2)
* [Entry 3: 2020-01-22, Second Info Updates](#id-section3)
* [Entry 4: 2020-01-27, Fisrt Paper Discussion](#id-section9)
* [Entry 5: 2020-01-29, Second Tutorial](#id-section10)
* [Entry 11: 2020-02-03, Second Paper Dicsussion](#id-section11)
* [Entry 18: 2020-02-05, Third Tutorial](#id-section18)
* [Entry 19: 2020-02-06, Thursday](#id-section19)
* [Entry 20: 2020-02-07, Friday](#id-section20)
* [Entry 21: 2020-02-10, Monday](#id-section21)
* [Entry 22: 2020-02-11, Tuesday](#id-section22)
* [Entry 23: 2020-02-12, Wednesday](#id-section23)
* [Entry 24: 2020-02-13, Thursday](#id-section24)
* [Entry 25: 2020-02-14, Friday](#id-section25)
* [Entry 26: 2020-02-17, Monday](#id-section26)
* [Entry 27: 2020-02-18, Tuesday](#id-section27)
* [Entry 28: 2020-02-19, Wednesday](#id-section28)
* [Entry 29: 2020-02-20, Thursday](#id-section29)
* [Entry 30: 2020-02-21, Friday](#id-section30)
* [Entry 31: 2020-02-24, Monday](#id-section31)
* [Entry 32: 2020-02-25, Tuesday](#id-section32)
* [Entry 33: 2020-02-26, Wednesday](#id-section33)
* [Entry 34: 2020-02-27, Thursday](#id-section34)
* [Entry 35: 2020-02-28, Friday](#id-section35)
* [Entry 36: 2020-03-02, Monday](#id-section36)
* [Entry 37: 2020-03-03, Tuesday](#id-section37)
* [Entry 38: 2020-03-04, Wednesday](#id-section38)
* [Entry 39: 2020-03-05, Thursday](#id-section39)
* [Entry 40: 2020-03-06, Friday](#id-section40)
* [Entry 41: 2020-03-09, Monday](#id-section41)
* [Entry 42: 2020-03-10, Tuesday](#id-section42)
* [Entry 43: 2020-03-11, Wednesday](#id-section43)
* [Entry 44: 2020-03-12, Thursday](#id-section44)
* [Entry 45: 2020-03-13, Friday](#id-section45)
* [Entry 46: 2020-03-16, Monday](#id-section46)
* [Entry 47: 2020-03-17, Tuesday](#id-section47)
* [Entry 48: 2020-03-18, Wednesday](#id-section48)
* [Entry 49: 2020-03-19, Thursday](#id-section49)
* [Entry 50: 2020-03-20, Friday](#id-section50)
* [Entry 51: 2020-03-23, Monday](#id-section51)
* [Entry 52: 2020-03-24, Tuesday](#id-section52)
* [Entry 53: 2020-03-25, Wednesday](#id-section53)
* [Entry 54: 2020-03-26, Thursday](#id-section54)
* [Entry 55: 2020-03-27, Friday](#id-section55)
* [Entry 56: 2020-03-30, Monday](#id-section56)
* [Entry 57: 2020-03-31, Tuesday](#id-section57)
* [Entry 58: 2020-04-01, Wednesday](#id-section58)
* [Entry 59: 2020-04-02, Thursday](#id-section59)
* [Entry 60: 2020-04-03, Friday](#id-section60)
* [Entry 61: 2020-04-06, Monday](#id-section61)
* [Entry 62: 2020-04-07, Tuesday](#id-section62)
* [Entry 63: 2020-04-08, Wednesday](#id-section63)
* [Entry 64: 2020-04-09, Thursday](#id-section64)
* [Entry 65: 2020-04-10, Friday](#id-section65)
* [Entry 66: 2020-04-13, Monday](#id-section66)
* [Entry 67: 2020-04-14, Tuesday](#id-section67)
* [Entry 68: 2020-04-15, Wednesday](#id-section68)
* [Entry 69: 2020-04-16, Thursday](#id-section69)
* [Entry 70: 2020-04-17, Friday](#id-section70)
* [Entry 71: 2020-04-20, Monday](#id-section71)
* [Entry 72: 2020-04-21, Tuesday](#id-section72)
* [Entry 73: 2020-04-22, Wednesday](#id-section73)
* [Entry 74: 2020-04-23, Thursday](#id-section74)
* [Entry 75: 2020-04-24, Friday](#id-section75)
* [Entry 76: 2020-04-27, Monday](#id-section76)
* [Entry 77: 2020-04-28, Tuesday](#id-section77)
* [Entry 78: 2020-04-29, Wednesday](#id-section78)
* [Entry 79: 2020-04-30, Thursday](#id-section79)
* [Entry 80: 2020-05-01, Friday](#id-section80)
* [Entry 81: 2020-05-04, Monday](#id-section81)
* [Entry 82: 2020-05-05, Tuesday](#id-section82)
* [Entry 83: 2020-05-06, Wednesday](#id-section83)
* [Entry 84: 2020-05-07, Thursday](#id-section84)
* [Entry 85: 2020-05-08, Friday](#id-section85)


------    
<div id='id-section1'/>   


### Entry 1: 2020-01-13, Introductory Class.   

#### Ecological Genomics:
* species interactions @ genomic look. 

* Study of genes with relation to organims interacting with the enviroment.

* Inherently about evolution. 

#### Population Genetics:
* 1920s-1930s

* Sewall Wright and R.A. Fisher

* Mathematical Theory:

1.  A1A2

2. p + q = 1 (allele frequencies adding to 1)

3.  HWE: 1 = p^2 + 2pq + q^2

4. Whright-Fisher population: infinitely large, cloased and unfragmented, random mating, no natural selection

* When N is finite and there are sampling effects, genetic drift occur.

* Island Model: Multiple populations, subdivided but connected through migration.

* Fst= (var(p))/(mean(p)(mean(q)))

#### Selective Sweeps:

* 1970s: allozymes

* 1980s: PCR, microsatelltes, DNA sequences using Sanger's method

#### Molecular Evolution:

* M. Kimura, important architect of moleculat evolution field. 

* Neutral Theory of Molecular Evolution.

* Probability of a new mutation 1/2N = mu and the the prbability of being fixed.

* dN/dS = (Nmutations/Nsites)/(Synonimous mutations/S sites) = w, if: 

1. w >> 1 w>2 there is positive selection

2. w = 1 neutral

3. w << 1 negative or purifying selections
  
* Tajima's D

1. pi = nucleotide diversity

2. S = number of segregating sites

3. Pairwise calculations. 

------    
<div id='id-section2'/>   


### Entry 2: 2020-01-15, First info-update class.   

#### To have in mind!

* Your Questions!

* Study organisms (Is there a reference genome? How big? Diploid, Hapliod, Polyploids?)

* Type of Sample (DNA, RNA, combination?)

* Library: biological library that has not been prepared.

1. Ex: Whole genome sequences, RNA Seq, Exom catcher, GBS, Amplicon Sequencing, Associated methods for epigenomics.

* Sequencing platform: Ilumina for short reads sequencing, Pac Bio and Nano probe for long reads.

#### Info updates about: 

1. Illumina, Long Reads sequences.

2. RNA Seq, GNS, Bisulfite sequencing, Exome Capture Sequencing.

------    
<div id='id-section3'/>   


### Entry 3: 2020-01-22, Second info-updates class.   

### My Info Update Presentation:

* Amplicon: Is a piece of DNA or RNA that is the source and/or product of amplification or replication events.

* In other words, is a segment of genetic material that undergoes simplification and contains replicated genetic material. Depending on what are you working with, amplicons can be direct repeat or inverted repeat, and can be linear or circular in structure. 

* Important: the term amplicon is often used interchangeably with PCR products.

* Amplicon sequences results in high coverage of a specific region, making amplicon sequences useful to detect variants at very low levels and frequencies. Also, it allows for multiplexing of samples.

* Useful for: genome targeting and detection  of hot-spot mutations, gene fusions, SNPs and in disciplines such as bacterial metagenomics or biodiversity assesments. 

### Steps:

1. Sample Selection: Environmental Samples. 

2. DNA or RNA extractions.

3. Library preparation:

3a. First PCR to amplify target genes with primers synthethized with a head sequence at the end.

3b. Optimal purification

3c. Second PCR to amplify products with primers that consist of the head sequence and include at the 5' end a library-specific barcode. 

3d. Purification with magnetic beads and stand.

3e. Once you get an equimlar amount of purified amplicons they are pooled.

4. Sequencing: Miseq Illumina Sequencing.


#### First in class coding:

[First coding tutorial](https://pespenilab.github.io/Ecological-Genomics/Tutorial/2020-01-22_Command_Line_Unix.html)

* Major calls: 

1. pwd : to know where do you live on the server

2. ~/ : same thing as typing /users/j/r/jruizaro

3. cd : change directory

4. mkdir : make directory

5. ll : to look what is inside directory

6. head : to look first rows of a document


------    
<div id='id-section9'/>   


### Entry 4: 2020-01-27, First paper discussion.   

[Temporal genomic contrasts reveal rapid evolutionary responses in an alpine mammal during recent climate change](https://journals.plos.org/plosgenetics/article?id=10.1371/journal.pgen.1008119)

* Take Home Message: Environmental changes inluences demographic patterns with quantifiable changes to selection and genetic drift. 

### Common Vocabulary

1. Population Genomics: selection and drift across the genome. 

* Drift: demographic - changes in abundance and in distribution, cause by environmental changes.

* Selection: changes in selection.

* In population genomics, linkages and interactions are studied, in population genetics not. 

2. Looking across the genome is challenging cause signals can be undistinguishable. We need to separate and differenctiate the results that demographhic changes and selection leave at the genome.

3. Genomic environment is an environment by itself.

4. Linkage disequilibrium: non-random association of alleles at two or more loci in a general population. 

5. We expect selection to act locality and on a particular site with LD, while we expect drift to act globally (affecting the mean, variance and the diversity).

6. Demographic history: 

* Ne = effective population size, size where a population behave as if it were a W-F population.

* Smaller population size, greater the drift.

* Migration = gene flow among and between populationg.

* Divergence time. 

* Coalescent theory: probability of two random alleles copies sharing a common ancestor or accommon gene in the past.

* 2Ne allele copies in a population of Ne individuals. 

* We need to sample enough to capture the most recent common ancestor,

* If a population is changing in size, that will affect the rate of coalescence appearance. You can use this know understand the behavior of the population (ex: bottleneck, expansion, etc.)

* If Ne split = divergence, first you will see a lot of colescence and back in time it will slow down. 

* In a bottleneck, you will lose many of the rare lineages and retain common ones.

* Site-frequency expectrum.

7. Positive Selection

------    
<div id='id-section10'/>   


### Entry 5: 2020-01-29, Second Tutorial: Working with batch scripts.   

[Second Tutorial](https://pespenilab.github.io/Ecological-Genomics/2020-01-29_PopGenomics_Day1.html)

* Species of Interest: Red spruce. Tree species that has undergoes through latitudinal and elevational migration. Now there are unconnected and fragmentated populations in the top of the appalachians going from New York to South Carolina. 

* In our dataset, we are working with genomic DNA of 110 mother trees from 23 populations. 

#### Pipeline:

1. .fastq.gz, we will visualize them using FastQC.

2. trimmed.fastq, using Trimmomatic.

3. Visualize the post trimming.

4. Map or alignment to the references genome.

5. Map (or align) to the reference genome using bwa. File: .sam.

6. post.processing using samtools and sambamba. Making a binary file: .bam.

7. remove drops, stats

[Second Tutorial](https://pespenilab.github.io/Ecological-Genomics/2020-01-29_PopGenomics_Day1.html) 

* Because files inside RS_ExomeSeq are .gz, to see the individual file we can no use head.We use zcat instead, to see it without uncomprresing it: (zcat NAME) | head -n 4. 

* We will see the barcodes indexes (pair) with a four line with a quality encoding score called the Q score. 

* Phred Quality Score: 10 or < 90 % accuracy 20 or > more than 99% accuracy: example p=10^(-38/10)= prob(of the machine misscalling the nucleotide on the loci) = 0.0038.

#### Population assigned to me: WA. 

#### Step:

1. vim in myscripts

2. press i

3. #!/bin/bash

4. cd ~/Ecological_Genomics_2020/myresults/ 

5. mkdir fastqc

6. for file /data/project_data/RS_ExomeSeq/fastq/edge_fastq/WA*fastq.gz

6.5. do

7. Output file in /fastqc: fastqc ${file} -o fastqc/

8. Closing the loop: done

9. Exit editing mode: ESC, shift :, wq fastqc.sh

* To move files aroud: mv filename dir/ 

10. permision: -rw-r--r-- without x means file is not executable, then chmod u+x fastqc.sh.

11. bash fastqc.sh

12. Push everything and pull it bacj to your local repository.

#### Now, we are going to trim.

1. Using Trimmomatic.

2. Download from data/scripts and save it into my scripts: cp trim_loop.sh ~/REPOSITORY/myscripts/

3. Open it in vi (filename) nad edit your population name.

4. bash trim_loop.sh


------    
<div id='id-section11'/>   


### Entry 6: 2020-02-03, Second Paper Dicsussion   

### Info-updated

### Genomic diversity - How to calculated and what it does it say?

* Effective population size = population size of an ideal population that follows W-F assumptions.

* Factors deviating Ne from N: 

1. Non-random mating

2. Selection

3. Reproductive Skew 

4. Non-constant population size

5. Unequal sex ratio (Segwall Wright) <= Ne = (4Nm x Nf)/(Nm + Nf)

6. Can be estimated by previous formulas, but there will be other variables (ex. mutations, migration, etc.) which will end up affecting Ne.

7. nucleotide diversity (or pi) = 4(Ne or net probability difference in nucleotide) u (or mutation rate with units of changes per site per generation)

* Nei's genetic distance - measurement of diversity

* Fst (population structure) = 1/(pi + 1) (as pi increases Fst decreases, menas population is more homogenous)

* or Fst = 1 - (Hs (or average H of subpop.)/Ht (or average H for whole pop.))

* Heterozygozity (H): condition of having two different alleles at a locus.

* Site Frequency Spectrum: distribution of the allele frequencies of a given set of loci (often SNPs) in a population or sample. 

* Tajima's D: statistic that compares the average number of pairwise differences with the number of segregating sites. If = 0, in equilibrium, factors are not influencing too much. < 0 (expansion, we will have more rare alleles or more rare mutations appearing) or > 0 (contraction).  

### Paper Discussion (Evolution of sociality in spiders leads to depleted genomic diversity at both population and species level)

Take Home Message: How the transition to sociology can affect Ne? Non-random mating, Reproductive Skew, Unequal sex ratio.

* Fst = 1 - (average pi of subpopulation/ total species pi)

* lower K = more geneflow

* Inbreeding coefficient: how much inbreeding is happening.

* Migration push TMRCA further back in the past. 

* Social = total pi = 0.00056, Ne = 21 x 10^3

* Subsocial = total pi = 0.00474, Ne = 178 x 10^3

* mutation rate = 6.65 x 10^-9


------    
<div id='id-section18'/>   


### Entry 7: 2020-02-05, Third Tutorial   

[Third Tutorial](https://pespenilab.github.io/Ecological-Genomics/2020-02-05_PopGenomics_Day2.html)

#### Mapping cleaned and trimmed reads against the reference genome

[congenie](http://congenie.org/) to look for availability of reference genomes.

* wget -> to get genomes from the internet

* We are going to use a subsetted of the reference genome to include just the contigs that contain one or more probes from our exon capture experiment. 



------    
<div id='id-section8'/>   


### Entry 8: 2020-01-22, Wednesday.   



------    
<div id='id-section9'/>   


### Entry 9: 2020-01-23, Thursday.   



------    
<div id='id-section10'/>   


### Entry 10: 2020-01-24, Friday.   



------    
<div id='id-section11'/>   


### Entry 11: 2020-01-27, Monday.   



------    
<div id='id-section12'/>   


### Entry 12: 2020-01-28, Tuesday.   



------    
<div id='id-section13'/>   


### Entry 13: 2020-01-29, Wednesday.   



------    
<div id='id-section14'/>   


### Entry 14: 2020-01-30, Thursday.   



------    
<div id='id-section15'/>   


### Entry 15: 2020-01-31, Friday.   



------    
<div id='id-section16'/>   


### Entry 16: 2020-02-03, Monday.   



------    
<div id='id-section17'/>   


### Entry 17: 2020-02-04, Tuesday.   



------    
<div id='id-section18'/>   


### Entry 18: 2020-02-05, Wednesday.   



------    
<div id='id-section19'/>   


### Entry 19: 2020-02-06, Thursday.   



------    
<div id='id-section20'/>   


### Entry 20: 2020-02-07, Friday.   



------    
<div id='id-section21'/>   


### Entry 21: 2020-02-10, Monday.   



------    
<div id='id-section22'/>   


### Entry 22: 2020-02-11, Tuesday.   



------    
<div id='id-section23'/>   


### Entry 23: 2020-02-12, Wednesday.   



------    
<div id='id-section24'/>   


### Entry 24: 2020-02-13, Thursday.   



------    
<div id='id-section25'/>   


### Entry 25: 2020-02-14, Friday.   



------    
<div id='id-section26'/>   


### Entry 26: 2020-02-17, Monday.   



------    
<div id='id-section27'/>   


### Entry 27: 2020-02-18, Tuesday.   



------    
<div id='id-section28'/>   


### Entry 28: 2020-02-19, Wednesday.   



------    
<div id='id-section29'/>   


### Entry 29: 2020-02-20, Thursday.   



------    
<div id='id-section30'/>   


### Entry 30: 2020-02-21, Friday.   



------    
<div id='id-section31'/>   


### Entry 31: 2020-02-24, Monday.   



------    
<div id='id-section32'/>   


### Entry 32: 2020-02-25, Tuesday.   



------    
<div id='id-section33'/>   


### Entry 33: 2020-02-26, Wednesday.   



------    
<div id='id-section34'/>   


### Entry 34: 2020-02-27, Thursday.   



------    
<div id='id-section35'/>   


### Entry 35: 2020-02-28, Friday.   



------    
<div id='id-section36'/>   


### Entry 36: 2020-03-02, Monday.   



------    
<div id='id-section37'/>   


### Entry 37: 2020-03-03, Tuesday.   



------    
<div id='id-section38'/>   


### Entry 38: 2020-03-04, Wednesday.   



------    
<div id='id-section39'/>   


### Entry 39: 2020-03-05, Thursday.   



------    
<div id='id-section40'/>   


### Entry 40: 2020-03-06, Friday.   



------    
<div id='id-section41'/>   


### Entry 41: 2020-03-09, Monday.   



------    
<div id='id-section42'/>   


### Entry 42: 2020-03-10, Tuesday.   



------    
<div id='id-section43'/>   


### Entry 43: 2020-03-11, Wednesday.   



------    
<div id='id-section44'/>   


### Entry 44: 2020-03-12, Thursday.   



------    
<div id='id-section45'/>   


### Entry 45: 2020-03-13, Friday.   



------    
<div id='id-section46'/>   


### Entry 46: 2020-03-16, Monday.   



------    
<div id='id-section47'/>   


### Entry 47: 2020-03-17, Tuesday.   



------    
<div id='id-section48'/>   


### Entry 48: 2020-03-18, Wednesday.   



------    
<div id='id-section49'/>   


### Entry 49: 2020-03-19, Thursday.   



------    
<div id='id-section50'/>   


### Entry 50: 2020-03-20, Friday.   



------    
<div id='id-section51'/>   


### Entry 51: 2020-03-23, Monday.   



------    
<div id='id-section52'/>   


### Entry 52: 2020-03-24, Tuesday.   



------    
<div id='id-section53'/>   


### Entry 53: 2020-03-25, Wednesday.   



------    
<div id='id-section54'/>   


### Entry 54: 2020-03-26, Thursday.   



------    
<div id='id-section55'/>   


### Entry 55: 2020-03-27, Friday.   



------    
<div id='id-section56'/>   


### Entry 56: 2020-03-30, Monday.   



------    
<div id='id-section57'/>   


### Entry 57: 2020-03-31, Tuesday.   



------    
<div id='id-section58'/>   


### Entry 58: 2020-04-01, Wednesday.   



------    
<div id='id-section59'/>   


### Entry 59: 2020-04-02, Thursday.   



------    
<div id='id-section60'/>   


### Entry 60: 2020-04-03, Friday.   



------    
<div id='id-section61'/>   


### Entry 61: 2020-04-06, Monday.   



------    
<div id='id-section62'/>   


### Entry 62: 2020-04-07, Tuesday.   



------    
<div id='id-section63'/>   


### Entry 63: 2020-04-08, Wednesday.   



------    
<div id='id-section64'/>   


### Entry 64: 2020-04-09, Thursday.   



------    
<div id='id-section65'/>   


### Entry 65: 2020-04-10, Friday.   



------    
<div id='id-section66'/>   


### Entry 66: 2020-04-13, Monday.   



------    
<div id='id-section67'/>   


### Entry 67: 2020-04-14, Tuesday.   



------    
<div id='id-section68'/>   


### Entry 68: 2020-04-15, Wednesday.   



------    
<div id='id-section69'/>   


### Entry 69: 2020-04-16, Thursday.   



------    
<div id='id-section70'/>   


### Entry 70: 2020-04-17, Friday.   



------    
<div id='id-section71'/>   


### Entry 71: 2020-04-20, Monday.   



------    
<div id='id-section72'/>   


### Entry 72: 2020-04-21, Tuesday.   



------    
<div id='id-section73'/>   


### Entry 73: 2020-04-22, Wednesday.   



------    
<div id='id-section74'/>   


### Entry 74: 2020-04-23, Thursday.   



------    
<div id='id-section75'/>   


### Entry 75: 2020-04-24, Friday.   



------    
<div id='id-section76'/>   


### Entry 76: 2020-04-27, Monday.   



------    
<div id='id-section77'/>   


### Entry 77: 2020-04-28, Tuesday.   



------    
<div id='id-section78'/>   


### Entry 78: 2020-04-29, Wednesday.   



------    
<div id='id-section79'/>   


### Entry 79: 2020-04-30, Thursday.   



------    
<div id='id-section80'/>   


### Entry 80: 2020-05-01, Friday.   



------    
<div id='id-section81'/>   


### Entry 81: 2020-05-04, Monday.   



------    
<div id='id-section82'/>   


### Entry 82: 2020-05-05, Tuesday.   



------    
<div id='id-section83'/>   


### Entry 83: 2020-05-06, Wednesday.   



------    
<div id='id-section84'/>   


### Entry 84: 2020-05-07, Thursday.   



------    
<div id='id-section85'/>   


### Entry 85: 2020-05-08, Friday.   
